source/extensions/clusters/dns_srv/dns_srv_cluster.cc

```diff
@@ -0,0 +1,351 @@
+  const auto& lb_endpoint_ = load_assignment_.endpoints()[0].lb_endpoints()[0];
+
+  PriorityStateManager priority_state_manager(*this, local_info_, nullptr, random_);
+  priority_state_manager.initializePriorityFor(locality_lb_endpoints);
+
+  // Only update the list of targets if at least one of them was resolved successfully
+  // this has a down-side: we can loose some of targets for some while but hopefully they'll get
+  // back next time
+  bool some_targets_resolved = false;
+
+  for (const auto& target : active_resolve_list_->getResolvedTargets()) {
+    // SRV query returns a number of instances (hostnames), but each hostname
+    // may potentially be resolved in a number of IP addresses
+    if (target->resolve_status_ != Network::DnsResolver::ResolutionStatus::Completed) {
+      ENVOY_LOG(debug, "IP resolution for target: '{}' has failed: {}",
+                target->srv_record_hostname_, target->resolve_status_details_);
+      continue;
+    }
+
+    some_targets_resolved = true;
+
+    for (auto& address : target->resolved_targets_) {
+      if (all_new_hosts.count(address->asString()) > 0) {
+        continue;
+      }
+
+      ENVOY_LOG(debug, "Endpoints size: {}", load_assignment_.endpoints().size());
+
+      auto host_or_status = HostImpl::create(
+          info_, target->srv_record_hostname_, address,
+          // TODO(zyfjeff): Created through metadata shared pool
+          std::make_shared<const envoy::config::core::v3::Metadata>(lb_endpoint_.metadata()),
+          std::make_shared<const envoy::config::core::v3::Metadata>(
+              locality_lb_endpoints.metadata()),
+          lb_endpoint_.load_balancing_weight().value(), locality_lb_endpoints.locality(),
+          lb_endpoint_.endpoint().health_check_config(), locality_lb_endpoints.priority(),
+          lb_endpoint_.health_status(), time_source_);
+
+      if (!host_or_status.ok()) {
+        info_->configUpdateStats().update_failure_.inc();
+
+        ENVOY_LOG(debug, "Failed to create host record for: '{}': code={}, message='{}'",
+                  target->srv_record_hostname_, host_or_status.status().raw_code(),
+                  host_or_status.status().message());
+        continue;
+      }
+
+      new_hosts.emplace_back(std::move(host_or_status.value()));
+      all_new_hosts.emplace(address->asString());
+
+      priority_state_manager.registerHostForPriority(new_hosts.back(), locality_lb_endpoints);
+    }
+
+    info_->configUpdateStats().update_success_.inc();
+  }
+
+  if (some_targets_resolved) {
+    HostVector all_hosts;
+    HostVector hosts_added;
+    HostVector hosts_removed;
+    HostVector hosts_;
+
+    if (updateDynamicHostList(new_hosts, hosts_, hosts_added, hosts_removed, all_hosts_,
+                              all_new_hosts)) {
+      // Update host map for current resolve target.
+      for (const auto& host : hosts_removed) {
+        all_hosts_.erase(host->address()->asString());
```

Comment:
Is it valid in SRV to have the same address with different priorities? If so, this map of IP address strings isn't expressive enough.